### PR TITLE
CLDC-2067 Bulk upload for lettings 23/24

### PR DESCRIPTION
**Pull request by asmega**
# Context

- https://digital.dclg.gov.uk/jira/browse/CLDC-2067
- This adds lettings 23/24 to buik upload

# Changes

- The new fields present at the end of the spreadsheet are not processed at the moment. This is because we are still waiting for the UPRN work to be completed which will add these new fields to the system
- Supports new lettings 23/24 template which has re-numbered all the fields
- This new template can be ordered arbitrarily assuming headers are present
- Despite being outside the crossover period i have forced this so user can choose which year to bulk upload for on deployed environments

### CODE REVIEW

File: app/services/bulk_upload/lettings/year2023/csv_parser.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+  def rows
+    @rows ||= CSV.parse(normalised_string, row_sep:)
+  end
+
+private
+
+  def default_field_numbers
+    [5, nil, nil, 15, 16, nil, 13, 40, 41, 42, 43, 46, 52, 56, 60, 64, 68, 72, 76, 47, 53, 57, 61, 65, 69, 73, 77, 51, 55, 59, 63, 67, 71, 75, 50, 54, 58, 62, 66, 70, 74, 78, 48, 49, 79, 81, 82, 123, 124, 122, 120, 102, 103, nil, 83, 84, 85, 86, 87, 88, 104, 109, 107, 108, 106, 100, 101, 105, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 126, 128, 129, 130, 131, 132, 127, 125, 133, 134, 33, 34, 35, 36, 37, 38, nil, 7, 8, 9, 28, 14, 32, 29, 30, 31, 26, 27, 25, 23, 24, nil, 1, 3, 2, 80, nil, 121, 44, 89, 98, 92, 95, 90, 91, 93, 94, 97, 96, 99, 10, 11, 12, 45, 39, 6, 4, 17, 18, 19, 20, 21, 22].map { |h| h.present? ? "field_#{h}" : "field_blank" }
+  end
```
Comment (kosiakkatrina) [inline app/services/bulk_upload/lettings/year2023/csv_parser.rb:53]: What do the nils/field_blank mean?

File: app/services/bulk_upload/lettings/year2023/csv_parser.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+  def field_numbers
+    @field_numbers ||= if with_headers?
+                         rows[row_offset - 1][col_offset..].map { |h| h.present? ? "field_#{h}" : "field_blank" }
+                       else
+                         default_field_numbers
+                       end
+  end
```
Comment (kosiakkatrina) [inline app/services/bulk_upload/lettings/year2023/csv_parser.rb:61]: Is there an equivalent of this for 2022?

File: spec/support/bulk_upload/log_to_csv.rb [hunk 3]
```diff
@@ -8,9 +8,49 @@ def initialize(log:, line_ending: "\n", col_offset: 1, overrides: {})
...
+      (row_prefix + row).flatten.join(",") + line_ending
+    else
+      (row_prefix + to_2023_row).flatten.join(",") + line_ending
+    end
+  end
+
+  def to_2023_row
+    to_2022_row + [
+      nil, # needstype,
+      nil, # location,
```
Comment (kosiakkatrina) [inline spec/support/bulk_upload/log_to_csv.rb]: Are we able to set these?

File: app/services/bulk_upload/lettings/year2023/csv_parser.rb (hunk 1 above)
Comment (asmega) [reply app/services/bulk_upload/lettings/year2023/csv_parser.rb:53]: yea these are blank columns where no data is present. due to backwards compatibility we keep these blank. these previously would of had some question in but later the question would have been removed.

File: app/services/bulk_upload/lettings/year2023/csv_parser.rb (hunk 2 above)
Comment (asmega) [reply app/services/bulk_upload/lettings/year2023/csv_parser.rb:61]: for 2022 all the columns are numbered. but some of those numbered fields are blank, so it works slightly differently

File: spec/support/bulk_upload/log_to_csv.rb (hunk 3 above)
Comment (asmega) [reply spec/support/bulk_upload/log_to_csv.rb]: unfortunately i can't add these yet. once the UPRN work is merged, which introduces these fields to the system we can start populating them

Comment (kosiakkatrina) [reply spec/support/bulk_upload/log_to_csv.rb]: but that doesn't include the needs type and location? Or am I missing something?

File: app/services/bulk_upload/lettings/year2023/csv_parser.rb (hunk 2 above)
Comment (kosiakkatrina) [reply app/services/bulk_upload/lettings/year2023/csv_parser.rb:61]: Do you mean 2022 columns can't be in random order?

Comment (asmega) [reply app/services/bulk_upload/lettings/year2023/csv_parser.rb:61]: i wasn't trying to say that but yes they can't be in a random order 😄

File: spec/support/bulk_upload/log_to_csv.rb (hunk 3 above)
Comment (asmega) [reply spec/support/bulk_upload/log_to_csv.rb]: oh yes i see now, good point. i can add those

**Review by kosiakkatrina — APPROVED** [review]

File: spec/support/bulk_upload/log_to_csv.rb (hunk 3 above)
Comment (asmega) [reply spec/support/bulk_upload/log_to_csv.rb]: just added

**Pull request merged**